controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -475,13 +468,52 @@ func generateKubeadmControlPlaneLabels(clusterName string) map[string]string {
 }
 
 // reconcileDelete handles KubeadmControlPlane deletion.
-func (r *KubeadmControlPlaneReconciler) reconcileDelete(_ context.Context, kcp *controlplanev1.KubeadmControlPlane, logger logr.Logger) (ctrl.Result, error) {
-	err := errors.New("Not Implemented")
+func (r *KubeadmControlPlaneReconciler) reconcileDelete(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, logger logr.Logger) (_ ctrl.Result, reterr error) {
+	// Fetch the Cluster.
+	cluster, err := util.GetOwnerCluster(ctx, r.Client, kcp.ObjectMeta)
+	if err != nil {
+		logger.Error(err, "Failed to retrieve owner Cluster from the API Server")
+		return ctrl.Result{}, err
+	}
+	if cluster == nil {
+		logger.Info("Cluster Controller has not yet set OwnerRef")
+		return ctrl.Result{}, nil
+	}
+	logger = logger.WithValues("cluster", cluster.Name)
 
+	// Fetch Machines
+	allMachines, err := util.GetMachinesForCluster(ctx, r.Client, cluster)
 	if err != nil {
-		logger.Error(err, "Not Implemented")
+		logger.Error(err, "Failed to get list of machines")
 		return ctrl.Result{}, err
 	}
+	ownedMachines := r.getOwnedMachines(kcp, allMachines)
+
+	// Always attempt to update status
+	defer func() {
+		if err := r.updateStatus(ctx, kcp, cluster); err != nil {
+			logger.Error(err, "Failed to update status")
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	// Verify that only control plane machines remain
+	if len(allMachines.Items) != len(ownedMachines.Items) {
+		err := errors.New("at least one machine is not owned by the control plane")
+		logger.Error(err, "Failed to delete the control plane")
+		return ctrl.Result{}, err
+	}
+
+	var errs []error
+	for i := range ownedMachines.Items {
+		m := &ownedMachines.Items[i]
+		if err := r.Client.Delete(ctx, m); err != nil && !apierrors.IsNotFound(err) {
```

Comment:
Do we need any sort of special ordering (e.g. newest to oldest) here or can we just go one by one?

Reply:
Good question.

No special ordering is needed if the control plane machines run control plane workloads only. These workloads are stateless, except for etcd. If the cluster admin wants a snapshot of etcd, they must create it before deleting the control plane. (Where is a good place to document this?)

If _other_ workloads run on control plane machines, they may need a special ordering in order to shut down gracefully. The current proposal does mention this use case explicitly. I think it's out of scope for this iteration; I can add it the proposal doc and mark it "out of scope."

Reply:
I'm fine with deferring handling non-control plane work loads on the control plane in this PR. Leave a comment in the code (a TODO maybe) and then open an issue/update the tracking issue.

Reply:
Thanks for feedback @chuckha (Sorry, looks like I forgot to tag you above).

I'll open an issue and leave a comment with a link.

Reply:
Filed https://github.com/kubernetes-sigs/cluster-api/issues/2064

Reply:
Added comment